src/drivers/kinetis/i2c.c

```diff
@@ -0,0 +1,310 @@
+                *ch->received_data++ = base->D;
+                break;
+            default:
+                /* simply perform a read */
+                *ch->received_data++ = base->D;
+                break;
+        }
+        /* decrement the number of reads */
+        ch->reads_ahead--;
+    }
+    /* handle writing */
+    else {
+        /* check if sequence end */
+        if(ch->seq == ch->seq_end) {
+            goto i2c_isr_stop;
+        }
+
+        /* check if NACK */
+        if(base->S & I2C_S_RXAK_MASK) {
+            goto i2c_isr_error;
+        }
+
+        /* check if we have a restart */
+        if((*ch->seq).opcode == I2C_RESTART) {
+            /* generate repeated start and ensure TX */
+            base->C1 |= (I2C_C1_RSTA_MASK | I2C_C1_TX_MASK);
+
+            /* address write post repeated start */
+            ch->seq++;
+            base->D = (*ch->seq).addr;
+        }
+        /* not a repeated start */
+        else {
+            /* handle a read */
+            if((*ch->seq).opcode == I2C_READ) {
+                /* set channel status to read */
+                ch->txrx = I2C_READING;
+
+                /* at least one read ahead */
+                ch->reads_ahead = 1;
+
+                /* parse the number of reads */
+                while(((ch->seq + ch->reads_ahead) < ch->seq_end) &&
+                        ((*(ch->seq + ch->reads_ahead)).opcode == I2C_READ)) {
+                    ch->reads_ahead++;
+                }
+
+                /* switch to rx */
+                base->C1 &= ~I2C_C1_TX_MASK;
+
+                if(ch->reads_ahead == 1) {
+                    /* do not ACK the final read */
+                    base->C1 |= I2C_C1_TXAK_MASK;
+                }
+                else {
+                    /* ACK every other read */
+                    base->C1 &= ~(I2C_C1_TXAK_MASK);
+                }
+
+                /* dummy read without incrementing received data pointer */
+                *ch->received_data = base->D;
+                ch->reads_ahead--;
+            }
+            /* handle a write */
+            else {
+                base->D = (*ch->seq).value;
+            }
+        }
+    }
+    ch->seq++;
+    goto i2c_isr_exit;
+
+i2c_isr_stop:
+    /* generate a STOP, switch to RX, disable interrupts */
+    base->C1 &= ~(I2C_C1_MST_MASK | I2C_C1_TXAK_MASK | I2C_C1_IICIE_MASK);
+
+    /* call the callback function on success if it exists */
+    if(ch->callback) {
+        (*ch->callback)(ch->args);
+    }
+
+    /* sest channel back to available */
+    ch->status = I2C_AVAILABLE;
+    goto i2c_isr_exit;
+
+i2c_isr_error:
+    /* generate a STOP and disable interrupts */
+    base->C1 &= ~(I2C_C1_MST_MASK | I2C_C1_IICIE_MASK);
+    ch->status = I2C_ERROR;
+    errno = -1;
+    goto i2c_isr_exit;
+
+i2c_isr_exit:
+    return errno;
+#ifdef __KINETIS_I2C_BLOCKING_ISR
+    asm("cpsie i");
+#endif
+}
+
+void I2C0_IRQHandler() {
+    i2c_irq_handler(0);
+}
```

Comment:
Where are the other handlers?

Reply:
Not using them right now so I omitted them. Should I add them in?

Reply:
Meh, I suppose we don't need them.  They're weakref'd anyway, so it would be easy to add them.

Reply:
Though on second thought - why is the ISR visible from the outside? Does it need to be? If not, probably hide it.